### PR TITLE
[ART-9330] [ART-8971] new embargo process

### PR DESCRIPTION
**Pull request by ashwindasr**
Ref https://issues.redhat.com/browse/ART-8080

The goal of this PR is to setup a new flow so that we can create embargoed nightlies (visible in the private release controller), so that QE and other stakeholders can test the payload well in advance of the embargo being lifted. Currently, we only start building after the embargo is lifted, and we immediately pin the build(s) in the payload and promote. Having the private builds earlier in the private release controller will let us follow the usual pattern of release, but with public nightlies. The flow can be summarized as follow:

- For STREAM:
    - Only public builds are synced to public image streams (thereby to public nightlies), but both public and private builds are synced to private nightlies. Since we rebase from openshift to openshift-priv before building from the latter, we make sure that the private builds always contain the latest public content.
  
- For non-STREAM:
    - A two key approach is adopted.
    - A parameter in build_sync has to be enabled as well. This param will set `--embargo-permit-ack`. (key 1)
    - If a private build is seen, it will fail viability, with code `EMBARGOED_CONTENT`. We need to permit this in the assembly definition (key 2)
     - Both the above condition has to be satisfied. Due to the nature of embargoes, the accidental release of embargoed content, before lift, is severe. Hence this level of gating is proposed.
     - gen-assembly can be run with private nightlies as params, instead of public nightlies.

### CODE REVIEW

File: doozer/doozerlib/cli/release_gen_payload.py
Comment (sosiouxme) [inline doozer/doozerlib/cli/release_gen_payload.py:783]: I think this can be simplified if at this point, we just don't consider the build private, and let it be included in the payload entries from the start, while noting that there are private entries, e.g.


```
                    # (note RHCOS builds can't be embargoes for now)
                    if self.runtime.assembly_type is not AssemblyTypes.STANDARD:
                        # Unless releasing, we do not want embargoed builds synced to public quay
                        continue
                    if self.cve_exclusion_ack:
                        # consciously releasing private content
                        self.includes_embargo = True
                        # probably create an AssemblyIssue somehow here to be permitted by the assembly?
                    else:
                        raise IOError(f"Embargoed build detected for payload tag {payload_tag_name} with "
                                                f"payload_entry: {payload_entry} \n --cve-exclusion-ack must be specified, "
                                                f"if this was intended")
```

Comment (ashwindasr) [reply doozer/doozerlib/cli/release_gen_payload.py]: > probably create an AssemblyIssue somehow here to be permitted by the assembly?
I was thinking we can have a flag in build sync?

File: artcommon/artcommonlib/assembly.py
Comment (thegreyd) [inline artcommon/artcommonlib/assembly.py:386]: I'm not sure what the utility of having it defined in assembly is .. or what we would be preventing.

File: artcommon/artcommonlib/assembly.py
Comment (jupierce) [inline artcommon/artcommonlib/assembly.py:387]: I would suggest a change here. `contains_embargoed_build` reads as if the stanza is asserting the payload does or does not contain a embargoed fix whereas I believe it is trying to imply "I know this payload contains something embargoed and we are ready to make it public". 
Permits are our typical method of capturing "I know this payload <is unusual for some reason>" semantics. So `code: UNEMBARGO` would make sense to me. 
Alternatively, a boolean like `acknowledge_unembargo` in the assembly, but I'd like to understand the rationale.

File: doozer/doozerlib/cli/release_gen_payload.py
Comment (jupierce) [inline doozer/doozerlib/cli/release_gen_payload.py:784]: Iterate through all results before throwing the error. Imagine if there were 2 embargoed images in the payload: one that was ready to be unembargoed and one that was not. You'd want the artist to know about both before making the decision to permit the unembargo.

**Review by jupierce — DISMISSED** [review]

File: artcommon/artcommonlib/assembly.py
Comment (sosiouxme) [reply artcommon/artcommonlib/assembly.py:387]: I thought when I looked previously that permits are not observed in build-sync, but it seems like they are (via `assembly_inspector.does_permit`). `release:gen-payload` could add an `AssemblyIssue` with code `EMBARGOED_CONTENT` whenever there's private content on a standard assembly, and then we could `permit` it in the assembly. I guess that would be nice to avoid adding a new attr.

File: doozer/doozerlib/cli/release_gen_payload.py
Comment (sosiouxme) [reply doozer/doozerlib/cli/release_gen_payload.py:784]: easy if it's just adding an `AssemblyIssue`. but this may require a little rearrangement; this sub doesn't currently contribute to assembly issues, and i don't recall where we actually check for viability. we want to make sure that's given a chance to evaluate and stop issues before we've actually copied any content out.

File: doozer/doozerlib/cli/release_gen_payload.py
Comment (sosiouxme) [inline doozer/doozerlib/cli/release_gen_payload.py:1938]: sounds like a boolean, should describe what it's returning
```suggestion
    def embargo_issues_for_payload(entries) -> List[AssemblyIssue]:
```

**Review by sosiouxme — CHANGES_REQUESTED** [review]
just some surface-level things

i'm not sure what to think about dividing into private/public sets of issues, let's talk about that

**Comment by ashwindasr** [comment]
Review issues fixed in https://github.com/openshift-eng/art-tools/pull/527/commits/d45438debbe81e6532393db3600f90cb59f154b4

**Comment by ashwindasr** [comment]
/unhold

Our automation has access to private nightlies as well

**Comment by ashwindasr** [comment]
/hold

Until testing is completed

File: doozer/doozerlib/cli/release_gen_payload.py
Comment (sosiouxme) [inline doozer/doozerlib/cli/release_gen_payload.py:795]: think you only wanted to filter embargoed ones here... this filters everything for `STREAM`

File: doozer/doozerlib/cli/release_gen_payload.py
Comment (sosiouxme) [inline doozer/doozerlib/cli/release_gen_payload.py:800]: nothing ever sets it, so how can it have any content?

i think you wanted to add the entries after the `if` "filters"

File: doozer/doozerlib/cli/release_gen_payload.py
Comment (sosiouxme) [inline doozer/doozerlib/cli/release_gen_payload.py:623]: i think we do not want to add these into the assembly issues list because it will just be the same issues as public, so we'll get duplicates (that's my memory looking at the logic this is replacing; and looks like issues only come from RHCOS at this point). the assembly issues, in addition to being checked for permits, are also added to annotations on the imagestream, so we don't want to bloat them any further.

**Review by sosiouxme — COMMENTED** [review]
sorry i'm out of time to review properly, will try to get back to it.
test method is pretty simple:
1. for an old release like 4.11, rebuild a p0 build as p1 (clone the distgit, change the release, brew container-build)
2. run build-sync to see if it correctly detects that this is an embargo
3. on a fork create a bogus release and build-sync it to see that it correctly handles the flag and permit being there or not.

File: doozer/tests/cli/test_gen_payload.py
Comment (sosiouxme) [inline doozer/tests/cli/test_gen_payload.py:214]: previously there was nothing in this method to inspect the output of `find_payload_entries`, which was just mocked with dummy responses; you added a call to `embargo_issues_for_payload` which actually look at those, so either need to come up with better responses, or mock that too.

File: doozer/doozerlib/cli/release_gen_payload.py
Comment (sosiouxme) [inline doozer/doozerlib/cli/release_gen_payload.py:742]:
```suggestion
            raise IOError(f"Embargoed components found: {set(permitted_non_acknowledged_embargoed_issues)}; "
                          "must specify --embargo-permit-ack flag to publish private content")
```

File: doozer/doozerlib/cli/release_gen_payload.py
Comment (sosiouxme) [inline doozer/doozerlib/cli/release_gen_payload.py:613]:
```suggestion
            entries_for_arch[arch] = {k: v for k, v in public_entries.items() if not v.is_under_embargo()}
```

File: doozer/doozerlib/cli/release_gen_payload.py
Comment (sosiouxme) [inline doozer/doozerlib/cli/release_gen_payload.py:806]: get rid of filtering at this point

**Review by sosiouxme — CHANGES_REQUESTED** [review]

**Comment by ashwindasr** [comment]
/unhold

Testing complete. Slack [thread](https://redhat-internal.slack.com/archives/GDBRP5YJH/p1715888032404899)

**Review by sosiouxme — CHANGES_REQUESTED** [review]
check my assertion that there would be a bug here...

File: doozer/doozerlib/cli/release_gen_payload.py
Comment (ashwindasr) [reply doozer/doozerlib/cli/release_gen_payload.py]: you're right, the original list was being processed again, instead of the filtered one

File: doozer/doozerlib/cli/release_gen_payload.py
Comment (thegreyd) [inline doozer/doozerlib/cli/release_gen_payload.py:628]: This should be public_entries i think since entries_for_arch is how the payload will be constructed
```suggestion
            entries_for_arch[arch] = public_entries
```

File: doozer/doozerlib/cli/release_gen_payload.py
Comment (thegreyd) [inline doozer/doozerlib/cli/release_gen_payload.py:640]: I think this will miss If there is an issue with a private build, so this should get us only the private build issues
```suggestion
            private_only_issues = set(private_payload_issues) - set(payload_issues)
            self.assembly_issues.extend(list(private_only_issues))
```

File: doozer/doozerlib/cli/release_gen_payload.py
Comment (thegreyd) [inline doozer/doozerlib/cli/release_gen_payload.py]: can we also rename this to `public_entries_for_arch`

File: doozer/doozerlib/cli/release_gen_payload.py
Comment (thegreyd) [inline doozer/doozerlib/cli/release_gen_payload.py:817]: let's rename the attr self.payload_entries_for_arch -> self.public_payload_entries_for_each
so it's crystal clear that these are supposed to be only public entries

**Comment by thegreyd** [comment]
Let's write a couple of unit tests, testing
- public and private builds are being returned in separate buckets
- when private build is in public bucket, the assembly issue code is being raised
- when assembly issue code is raised, we complain if ack flag is not found (and a case testing vice versa)

File: doozer/doozerlib/cli/release_gen_payload.py
Comment (sosiouxme) [inline doozer/doozerlib/cli/release_gen_payload.py:642]: i debated whether it made sense to add these at all. seems like it could be confusing if there were any and they showed up on the public payloads. OTOH they would show issues that could become public if the payload were promoted so... i guess it makes sense to include them.

**Review by sosiouxme — APPROVED** [review]
thanks, looks clearer. now. sid has some requests but you have a thumbs up from me.

**Comment by sosiouxme** [comment]
/hold
in case you want to address sid's requests before merging...
or that could be another PR :shrug:

**Comment by ashwindasr** [comment]
/hold

Until the team decides when to merge this to prod

**Comment by ashwindasr** [comment]
/unhold

due to a emergency

**Pull request merged**
